[DOCS] EMT Hypervisor Docs (#512)
open-edge-platform/edge-microvisor-toolkit: docs/developer-guide/get-started/deployment/emt-vm-host-ubuntu-guest.md

# Hosting Windows 11 Guest VM under EMT Host

## Install and Setup Ubuntu VM

### Prerequisites

1. Installation scripts

   Download the required installation scripts from the release package [RPL-S_RPL-SR_KVM_MultiOS.zip](https://github.com/intel-innersource/virtualization.multios.kvm.scripts/archive/refs/tags/rpls_sriov_kvm_multios_emt-3.1_ww2525.zip)

> 

Reviewer: ```suggestion
   Download the required installation scripts from the [RPL-S_RPL-SR_KVM_MultiOS.zip](https://github.com/intel-innersource/virtualization.multios.kvm.scripts/archive/refs/tags/rpls_sriov_kvm_multios_emt-3.1_ww2525.zip) release package.

```

   Run the following

   ```bash
   # Change to work directory
   $ cd ~

   # Copy files
   $ cp <source path>/virtualization.multios.kvm.scripts-rpls_sriov_kvm_multios_emt-3.1_ww2525.zip .

   # Extract script files

>    $ unzip -jo virtualization.multios.kvm.scripts-rpls_sriov_kvm_multios_emt-3.1_ww2525.zip

Reviewer: ```suggestion
   Run the following command:

   ```bash
   # Change to work directory
   cd ~

   # Copy files
   cp <source path>/virtualization.multios.kvm.scripts-rpls_sriov_kvm_multios_emt-3.1_ww2525.zip .

   # Extract script files
   unzip -jo virtualization.multios.kvm.scripts-rpls_sriov_kvm_multios_emt-3.1_ww2525.zip
```

   ```

2. Ubuntu ISO image

   Download Ubuntu 24.04.2 LTS image from the [link](https://cdimage.ubuntu.com/releases/jammy/release/inteliot/ubuntu-22.04-desktop-amd64+intel-iot.iso)

> 

Reviewer: ```suggestion
   [Download Ubuntu 24.04.2 LTS image](https://cdimage.ubuntu.com/releases/jammy/release/inteliot/ubuntu-22.04-desktop-amd64+intel-iot.iso)
```

#### Prepare Ubuntu VM image

> 1. Create symlink

Reviewer: ```suggestion
1. Create a symlink.
```

```bash

> $ ln -s ubuntu-22.04-desktop-amd64+intel-iot.iso ubuntu.iso

Reviewer: ```suggestion
ln -s ubuntu-22.04-desktop-amd64+intel-iot.iso ubuntu.iso
```

```

> 2. Create empty Ubuntu image file

Reviewer: ```suggestion
2. Create an empty Ubuntu image file.
```

```bash
$ cd /home/$USER/

> $ qemu-img create -f qcow2 ubuntu.qcow2 64G

Reviewer: ```suggestion
cd /home/$USER/
qemu-img create -f qcow2 ubuntu.qcow2 64G
```

```

> 3. Prepare OVMF Files

Reviewer: ```suggestion
3. Prepare OVMF files.
```

```bash
$ cd /home/$USER/
$ ln -sf ovmf/OVMF_CODE.fd OVMF_CODE.fd

> $ cp ovmf/OVMF_VARS.fd OVMF_VARS_ubuntu.fd

Reviewer: ```suggestion
cd /home/$USER/
ln -sf ovmf/OVMF_CODE.fd OVMF_CODE.fd
cp ovmf/OVMF_VARS.fd OVMF_VARS_ubuntu.fd
```

```

4. Create a separate copy of OVMF for Ubuntu VM use:

```bash
# Make a copy of OVMF for Ubuntu guest
$ ln -sf ovmf/OVMF_CODE.fd OVMF_CODE.fd

> $ cp ovmf/OVMF_VARS.fd OVMF_VARS_ubuntu.fd

Reviewer: ```suggestion
# Make a copy of OVMF for Ubuntu guest
ln -sf ovmf/OVMF_CODE.fd OVMF_CODE.fd
cp ovmf/OVMF_VARS.fd OVMF_VARS_ubuntu.fd
```

```

#### Install Ubuntu VM Image

1. Run `install_ubuntu.sh` to start Ubuntu guest VM installation.

   ```bash
   # Start guest VM to install Ubuntu

>    $ sudo ./install_ubuntu.sh

Reviewer: ```suggestion
   sudo ./install_ubuntu.sh
```

   ```

>    > **Note:** If guest VM enters UEFI shell instead of Ubuntu, please type the following in EFI shell:

Reviewer: ```suggestion
   > **Note:** If the guest VM enters a UEFI shell instead of Ubuntu, run the following commands in the EFI shell:
```

   ```bash
   fs0:
   cd efi/boot
   grubx64.efi
   ```

>    Then press enter and continue boot to Ubuntu as shown below.

Reviewer: ```suggestion
   Then, press the enter key and continue booting to Ubuntu, as shown below.
```

   ![Ubuntu Boot Screen](./assets/emt-vm-host-ubuntu/install.png)

> 2. Run Ubuntu OS installation to install the guest OS.

Reviewer: ```suggestion
2. Install Ubuntu for the guest OS.
```

   ![Ubuntu Install UI](./assets/emt-vm-host-ubuntu/image2021-12-14_14-10-9.png)

>    > **Note:** At the end of Ubuntu installation, the installer will ask you to remove the installation disk. Just press 'Enter' to reboot to the newly installed Ubuntu guest OS.

Reviewer: ```suggestion
   > **Note:** At the end of Ubuntu installation, you will be prompted to remove the installation disk. Just press 'Enter' to reboot to the guest OS.
```

### Configure Ubuntu Guest VM

1. Open a terminal within the Ubuntu guest VM.

> 2. Run the command shown below to upgrade Ubuntu software to the latest in the guest VM.

Reviewer: ```suggestion
2. Run the command shown below to upgrade Ubuntu software.
```

   ```bash
   # Upgrade Ubuntu software
   $ sudo apt -y update

>    $ sudo apt -y upgrade

Reviewer: ```suggestion
   sudo apt -y update
   sudo apt -y upgrade
```

   ```

   > **Note:** If operating behind a corporate firewall, setup proxy settings as required.

> 3. Install IOTG kernel and configure Ubuntu Guest OS

Reviewer: ```suggestion
3. Install IOTG kernel and configure the Ubuntu guest OS.
```

   Copy the following files and directories from the `/home/$USER` directory of the host OS to the `/home/$USER/` directory of the guest OS:

   - lts2024-iotg-kernel-rel.tar.gz
   - packages directory.

>    - sriov_install directory.

Reviewer: ```suggestion
   - lts2024-iotg-kernel-rel.tar.gz
   - the `packages` directory
   - the `sriov_install` directory
```

   Boot into the Ubuntu guest image:

   ```bash
   # Change to work directory
   $ cd ~

   # Copy install files from host into /home/$USER
   # Format:
   # scp -r <host_user>@<host_ip>:<host_source_dir>{file1,
   # file2,
   # file3,
   # dir1,
   # dir2} <guest_target_dir>
   # where,
   # <host_user>: the username of your host machine.
   # <host_ip>: the IP address of your host machine.
   # <host_source_dir>: source directory on the host
   # <guest_target_dir>: target directory on the guest
   $ scp -r <host_user>@<host_ip>:/home/<host_user>/{lts2024-iotg-kernel-rel.tar.gz,virtualization.multios.kvm.scripts-rpls_sriov_kvm_multios_emt-3.1_ww2525.zip,packages,sriov_install} /home/$USER

   # Extract script files
   $ unzip -jo virtualization.multios.kvm.scripts-rpls_sriov_kvm_multios_emt-3.1_ww2525.zip
   $ unzip sriov_patches.zip

   # Perform kernel setup
   # This will install kernel and firmware, and update grub
   $ sudo ./sriov_setup_kernel.sh

   # Reboot the guest

>    $ sudo reboot

Reviewer: ```suggestion
   cd ~

   # Copy install files from host into /home/$USER
   # Format:
   # scp -r <host_user>@<host_ip>:<host_source_dir>{file1,
   # file2,
   # file3,
   # dir1,
   # dir2} <guest_target_dir>
   # where,
   # <host_user>: the username of your host machine.
   # <host_ip>: the IP address of your host machine.
   # <host_source_dir>: source directory on the host
   # <guest_target_dir>: target directory on the guest
   scp -r <host_user>@<host_ip>:/home/<host_user>/{lts2024-iotg-kernel-rel.tar.gz,virtualization.multios.kvm.scripts-rpls_sriov_kvm_multios_emt-3.1_ww2525.zip,packages,sriov_install} /home/$USER

   # Extract script files
   unzip -jo virtualization.multios.kvm.scripts-rpls_sriov_kvm_multios_emt-3.1_ww2525.zip
   unzip sriov_patches.zip

   # Perform kernel setup
   # This will install kernel and firmware, and update grub
   sudo ./sriov_setup_kernel.sh

   # Reboot the guest
   sudo reboot
```

   ```

9. Reboot into the guest:

   ```bash
   # after reboot check the kernel version
   $ uname -r
   6.12.32-lts2024-iotg

   # Update and configure the guest

>    $ sudo ./configure_ubuntu_guest.sh

Reviewer: ```suggestion
   uname -r
   6.12.32-lts2024-iotg

   # Update and configure the guest
   sudo ./configure_ubuntu_guest.sh
```

   ```

>    > **IMPORTANT:** Please shutdown the guest properly once the installation has completed!

Reviewer: ```suggestion
   > **IMPORTANT:** Shutdown the guest OS properly once the installation has completed!
```

   The ubuntu.qcow2 image is now ready to use.

## Configuration Options for Running Ubuntu VM

### Changing Guest VM Memory and Number of CPUs

> The default launch command without any parameters is for 2 cores and 2G RAM. You can change that with startup parameters.

Reviewer: ```suggestion
The default launch command without any parameters is for 2 cores and 2GB of RAM. You can change that with startup parameters.
```

> Example guest start configuration for 4 cores, 4G RAM:

Reviewer: ```suggestion
Below is an example of guest start configuration for 4 cores and 4GB of RAM:
```

```bash
# Add -m option to specify 4G of memory
# Add -c option to specify 4 cpu cores for guest VM

> $ sudo -E ./start_ubuntu.sh -m 4G -c 4

Reviewer: ```suggestion
sudo -E ./start_ubuntu.sh -m 4G -c 4
```

```

### Enabling USB Devices in Guest VM

For Ubuntu guest VMs, USB devices can be setup in two ways:

> 1. Passthrough of All USB Host Devices

Reviewer: ```suggestion
1. Passthrough of all USB host devices.
```

> USB host passthrough parameter option can be added in the launch command to passthrough all USB devices on the USB host.

Reviewer: ```suggestion
The USB host passthrough parameter can be added in the launch command to passthrough all USB devices on the USB host.
```

> Add an additional parameter to the Guest VM launch command:

Reviewer: ```suggestion
Specify an additional parameter to the guest VM launch command:
```

```bash
# Note: all connected USB devices will be passthrough to the
guest VM with USB host passthrough option

> $ sudo -E ./start_ubuntu.sh --passthrough-pci-usb

Reviewer: ```suggestion
# guest VM with USB host passthrough option
sudo -E ./start_ubuntu.sh --passthrough-pci-usb
```

```

> 2. Passthrough of Selected USB Devices

Reviewer: ```suggestion
2. Passthrough of selected USB devices.
```

An external command option can be used to passthrough only a few selected USB devices.

> Retrieve the vendorid and productid of USB device. In this example, ‘046d’ is vendor

Reviewer: ```suggestion
Retrieve the "vendorid" and "productid" of the USB device. In this example, `046d` is a vendor
```

> ID, ‘c06a’ is product ID.

Reviewer: ```suggestion
ID, `c06a` is a product ID.
```

```bash

> # On target terminal.

Reviewer: ```suggestion
# In the target terminal.
```

> $ lsusb

Reviewer: ````suggestion
lsusb
```
````

> Bus 004 Device 003: ID 046d:c06a Logitech, Inc. USB Optical

Reviewer: ````suggestion

Displays the following information:
```text
Bus 004 Device 003: ID 046d:c06a Logitech, Inc. USB Optical
````

Mouse
```

> Add an additonal parameter to the Guest VM launch command:

Reviewer: ```suggestion
Specify an additonal parameter to the guest VM launch command:
```

```bash
# Add extra command when start guest

> $ sudo -E ./start_ubuntu.sh -e "-device usb-host,vendorid=0x046d,productid=0xc06a"

Reviewer: ```suggestion
# Add extra command when starting the guest OS
sudo -E ./start_ubuntu.sh -e "-device usb-host,vendorid=0x046d,productid=0xc06a"
```

```

> > **Note:** A passthrough device option can only be used once because a device can be passthrough to only 1 guest VM at a time.

Reviewer: ```suggestion
> **Note:** The passthrough device option can only be used once because a device can be passthrough to only one guest VM at a time.
```

### Enabling PCIe Wi-Fi Adapter Device in Guest VM

> For Ubuntu guest VMs, PCI Wi-Fi device passthrough can be setup by adding `--passthrough-pci-wifi` parameter to guest VM launch command:

Reviewer: ```suggestion
For Ubuntu guest VMs, PCI Wi-Fi device passthrough can be setup by adding the `--passthrough-pci-wifi` parameter to the guest VM launch command:
```

```bash
# Add --passthrough-pci-wifi for passing through Wifi adapter

> $ sudo -E ./start_ubuntu.sh --passthrough-pci-wifi

Reviewer: ```suggestion
sudo -E ./start_ubuntu.sh --passthrough-pci-wifi
```
